File: ROquiz/src/model/Quiz.java
package model;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;

public class Quiz {
	private List<Question> quiz;
	private int[] answers;
	
	public Quiz(List<Question> questions, int qNum)
	{
		this.quiz = new ArrayList<Question>();
		
		Collections.shuffle(questions);
		for(int i = 0; i < qNum; i++)
		{
			this.quiz.add(questions.get(i));
		}
		answers = new int[qNum];
		for(int i : answers)
			i = -1;
	}
	
	public List<Question> getQuiz() {return this.quiz;}
	public int[] getAnswers() {return this.answers;}
	
	public void setAnswer(int index, int value)
	{
		/*if(value < 1 || value > 5)
			throw new IllegalArgumentException();*/
		this.answers[index] = value;
	}
	
	public void resetQuiz(List<Question> questions, int qNum)
	{
		this.quiz = new ArrayList<Question>();
		Collections.shuffle(questions);
		for(int i = 0; i < qNum; i++)
		{
			this.quiz.add(questions.get(i));
		}
		answers = new int[qNum];
		for(int i : answers)
			i = -1;
	}
}

File: app-mobile/flutter_application/lib/model/Quiz.dart
import 'package:roquiz/model/Answer.dart';
import 'package:roquiz/model/Question.dart';

class Quiz {
  List<Question> questions = [];
  List<Answer> answers = []; // user answers (selected or unselected)
  int givenAnswers = 0; // total given answers
  int correctAnswer = 0; // total correct answers

  /*Quiz(List<Question> questions, int qNum, bool shuffleAnswers) {
    _resetQuiz(questions, qNum, shuffleAnswers);
  }*/

  void resetQuiz(List<Question> questions, int qNum, bool shuffleAnswers) {
    this.questions = [];
    answers = [];

    // reset Questions
    questions.shuffle();
    for (int i = 0; i < qNum; i++) {
      if (shuffleAnswers) questions[i].shuffleAnswers();

      this.questions.add(questions[i]);
      answers.add(Answer.NONE);
    }

    // reset Answer counters
    givenAnswers = 0;
    correctAnswer = 0;
  }

  void setAnswerFromInt(int index, int value) {
    if (answers[index] == Answer.NONE) {
      givenAnswers++;
    }

    answers[index] = Answer.values[value];
  }

  void setAnswerFromEnum(int index, Answer value) {
    if (answers[index] == Answer.NONE) {
      givenAnswers++;
    }
    answers[index] = value;
  }

  double getQuizGrade() {
    if (questions.isEmpty) return 0;
    double percentage = correctAnswer / questions.length;
    return 18 + percentage * (30 - 18);
  }

  double calculateFinalGrade(double writtenGrade) {
    double quizGrade = getQuizGrade();
    return ((writtenGrade * 2) + quizGrade) / 3;
  }

}
